keps/sig-node/4817-resource-claim-device-status/README.md

```diff
@@ -0,0 +1,618 @@
+management of resources.
+
+### Goals
+
+* Allow arbitrary, driver-specific information to be reported from the DRA
+  drivers for each allocated device in a ResourceClaim.
+* Establish a standardization for device status (e.g. Network Devices status).
+* Enable 3rd party implementation of new functionalities based on the Device
+  Status (e.g. Secondary Network Services if the IPs of a network device are
+reported).
+
+### Non-Goals
+
+* Implement new functionalities based on the Device Status (e.g. Kubernetes
+  EndpointSlice Controller supporting IPs reported by networking DRA drivers).
+* Modifying the Resource Claim workflow to include the device status report.
+
+## Proposal
+
+### API - ResourceClaim.Status
+
+The API changes define a new `DeviceStatuses` field in the existing
+`ResourceClaimStatus` struct. `DeviceStatuses` is a slice of a new struct
+`AllocatedDeviceStatus` which holds device specific information. 
+
+A device, identified by `<driver name>/<pool name>/<device name>` can be
+represented only once in the `DeviceStatuses` slice and will also mention which
+request caused the device to be allocated. The state and characteristics of the
+device are reported in the `Conditions`, representing the operational state of
+the device and in the `DeviceInfo`, an arbitrary data slice representing device
+specific characteristics. Additionally, for networking devices, a field
+`NetworkDeviceInfo` can be used to report the IPs, the MAC address and the
+interface name.
+
+`DeviceInfo` being a slice of arbitrary data allows the DRA Driver to store
+device specific data in different formats. For example, a Network Device being
+configured by via a CNI plugin could get its `DeviceInfo` filled with the CNI
+result for troubleshooting purpose and with a Network result in a modern
+standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
+controllers.
+
+For each device, if required, the DRA Driver processing the device allocation
+can report the status of it in the `Status.DeviceStatuses` of the ResourceClaim
+by using the Kubernetes API.
+
+```golang
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the result of that was.
+type ResourceClaimStatus struct {
+    ...
+
+    // DeviceStatuses contains the status of each device allocated for this
+    // claim, as reported by the driver. This can include driver-specific
+    // information. Entries are owned by their respective drivers.
+    //
+    // +optional
+    // +listType=map
+    // +listMapKey=driver
+    // +listMapKey=device
+    // +listMapKey=pool
+    DeviceStatuses []AllocatedDeviceStatus `json:"deviceStatuses,omitempty" protobuf:"bytes,4,opt,name=deviceStatuses"`
+}
+
+
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
```

Comment:
See also https://github.com/kubernetes/enhancements/pull/4861#discussion_r1770979535

Reply:
Regarding "not be allocated because it failed or something and we want to report back": if a device cannot be allocated for a claim, then there is no reason to publish anything about it in the claim. Otherwise that same information would have to be added to all claims which potentially might have used such a device.

Reply:
Seems we are not differentiating between "reserved by the scheduler" and "actuated by the driver" and are using "allocated" for both? We should probably differentiate.

Reply:
To use if I understand it, `AllocatedDeviceStatus` here means "actuated by the driver" , right? since the driver is the one that populates the information ...

Reply:
Yes.